### DISCUSSION TITLE
Nullable value types and switch expression

### DISCUSSION THREAD

**Discussion by Dersei** [dotnet/csharplang#2387] [Q&A] [answered]
I was playing with C# 8.0 in Visual Studio 2019 Preview and when I was testing switch expression I wrote something similar to this

```csharp
int? test = 0;
bool? tested = test switch
{
    0 => false,
    1 => true,
     null => null,
     _ => null
};
```

This results in error
`
CS0037	Cannot convert null to 'bool' because it is a non-nullable value type
`

It's easy to fix by casting any of result values to bool?, however I would like to know if there is any reason why this works in this way?

**Comment by YairHalberstadt** [comment]
expressions in C# are almost never target typed, so declaring tested as a `bool?` makes no difference to the type of the switch.

There are various proposals to improve this behaviour for the conditional operator. The same proposals could probably be applied here.

**Comment by ufcpp** [comment]
> The same proposals could probably be applied here.

https://github.com/dotnet/csharplang/issues/450.

**Comment by DavidArno** [comment]
~~This issue is now covered by the championed proposal, #2389, and so can be closed.~~

**Comment by Thaina** [comment]
@DavidArno Not sure that #2389 already covered nullable. It seem only mention the base type

**Comment by yaakov-h** [comment]
#2389 covers the target type, which is the exact case of the OP.

I wish it could cover nullable without target-typing (e.g. `var tested = ...`), but that's not quite the use case here.

**Comment by DavidArno** [comment]
@yaakov-h,

I'd assume #2389 should cover `var tested = …`, because it shouldn't be looking at the type of `tested` to determine the type of the expression. My assumption was that it would take the common type for `true`, `false` and `null`, which is `bool?`. I'll add something to that issue to seek clarification on this point.

**Comment by DavidArno** [comment]
@yaakov-h, so you were right. #2389 wouldn't work for `var tested = …`, so I've lost interest in it.

**Comment by gafter** [comment, marked as answer]
This would be fixed by either https://github.com/dotnet/csharplang/issues/2389 or https://github.com/dotnet/csharplang/issues/33.  It is likely that one or both of these will be included in C# 8.
